templates/theories/linear_algebra.md

```diff
@@ -0,0 +1,106 @@
+For example, the adjacency matrix of a graph could be indexed over the nodes in that graph.
+If you want to specify the dimensions of a matrix as natural numbers `m n : ℕ`, you can use `fin m` and `fin n` as index types.
+
+A matrix is constructed by giving the map from indices to entries: `(λ (i : m) (j : n), (_ : α)) : matrix m n α`.
+For matrices indexed by natural numbers, you can also use the notation defined in `data.matrix.notation`: `![![a, b, c], ![b, c, d]] : matrix (fin 2) (fin 3) α`.
+To get an entry of the matrix `M : matrix m n α` at row `i : m` and column `j : n`,
+you can apply `M` to the indices: `M i j : α`.
+Lemmas about the entries of a matrix typically end in `_val`: `add_val M N i j : (M + N) i j = M i j + N i j`.
+
+Matrix multiplication and transpose have notation that is made available by the command `open_locale matrix`.
+
+When working with matrices, a *vector* means a function `m → α` for an arbitrary `fintype` `m`.
+These have a module (or vector space) structure defined in `algebra.pi_instances`
+consisting of pointwise addition and multiplication.
+The distinction between row and column vectors is only made by the choice of function.
+For example, `mul_vec M v` multiplies a matrix with a column vector `v : m → α` and `vec_mul v M` multiplies a row vector `v : m → α` with a matrix.
+
+Permutation matrices are defined in `data.matrix.pequiv`.
+
+The determinant of a matrix is defined in `linear_algebra.determinant`.
+
+The adjugate and for nonsingular matrices, the inverse is defined in `linear_algebra.nonsingular_inverse`.
+
+The type `special_linear_group m R` is the group of `m` by `m` matrices with determinant `1`,
+and is defined in `linear_algebra.special_linear_group`.
+
+### linear maps and equivalences ###
+
+#### linear_algebra.basic ####
+
+The type `M →[R]ₗ M₂`, or `linear_map R M M₂`, represents `R`-linear maps from the `R`-module `M` to `M₂`.
+These are defined by their action on elements of `M`.
+The type `M ≃[R]ₗ M₂`, or `linear_equiv R M M₂`, is the type of invertible `R`-linear maps from `M` to `M₂`.
+
+The equivalence between matrices and linear maps is formalised in `linear_algebra.matrix`.
+`linear_equiv_matrix'` shows that `matrix.mul_vec` is a linear equivalence between `matrix m n R` and `(n → R) →[R]ₗ (m → R)`.
+The difference between matrices and linear maps is that matrices are in their essence an array of entries
+(which incidentally allows actions such as `matrix.mul_vec`),
+while linear maps are in their essence an action on vectors
+(which incidentally can be represented by a matrix if we have a finite basis).
+If you want to do computations, a matrix is a better choice.
+If you want to do proofs without computations, a linear map is a better choice.
+
+The type `general_linear_group R M` is the group of invertible `R`-linear maps from `M` to itself.
+`general_linear_equiv R M` is the equivalence between `general_linear_group` and `M ≃[R]ₗ M`.
```

Comment:
Why do we have two different objects here? Why isn't `general_linear_group R M` defined as `M ≃[R]ₗ M`?

Reply:
The practical difference in seems to be only in how the various proofs are bundled. The reason for this I don't know - I'll ask on Zulip.